specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,501 @@
+[usdt]: https://forum.openzeppelin.com/t/usdt-locked-in-a-contract-that-has-a-withdraw-token-function/32154/2
+
+## Configuring the Gas Paying Token
+
+The gas paying token is set within the L1 `SystemConfig` smart contract. This allows for easy access to the
+information required to know if an OP Stack chain is configured to use a custom gas paying token. The gas paying
+token is set during initialization and cannot be modified by the `SystemConfig` bytecode. Since the `SystemConfig`
+is proxied, it is always possible to modify the storage slot that holds the gas paying token address directly
+during an upgrade. It is assumed that the chain operator will not modify the gas paying token address unless they
+specifically decide to do so and appropriately handle the consequences of the action.
+
+The gas paying token address is network specific configuration, therefore it MUST be set in storage and not
+as an immutable. This ensures that the same contract bytecode can be used by multiple OP Stack chains.
+
+If the address in the `GAS_PAYING_TOKEN_SLOT` slot for `SystemConfig` is `address(0)`, the system is configured to
+use `ether` as the gas paying token, and the getter for the token returns `ETHER_TOKEN_ADDRESS`. If the address in
+the `GAS_PAYING_TOKEN_SLOT` slot for `SystemConfig` is not `address(0)`, the system is configured to use a custom
+gas paying token, and the getter returns the address in the slot.
+
+```mermaid
+---
+title: Custom Gas Token Configuration
+---
+flowchart LR
+    subgraph Layer One
+        OP[OptimismPortal]
+        subgraph SystemConfig
+          TA1[Token Address]
+        end
+        CO[Chain Operator] -->|"initialize()"| SystemConfig
+        SystemConfig -->|"setGasPayingToken()"| OP
+    end
+
+    subgraph Layer Two
+        subgraph L1Block
+          TA2[Token Address]
+        end
+    end
+    OP -->|"setGasPayingToken()"| L1Block
+```
+
+## Contract Modifications
+
+### IGasToken Interface
+
+This interface encapsulates the shared interface.
+
+```solidity
+interface IGasToken {
+    /// @notice Getter for the ERC20 token address that is used to pay for gas and its decimals.
+    function gasPayingToken() external view returns (address, uint8);
+    /// @notice Returns the gas token name.
+    function gasPayingTokenName() external view returns (string memory);
+    /// @notice Returns the gas token symbol.
+    function gasPayingTokenSymbol() external view returns (string memory);
+    /// @notice Returns true if the network uses a custom gas token.
+    function isCustomGasToken() external view returns (bool);
+}
+```
+
+If custom gas token is not used, then `gasPayingToken()` should return `(ETHER_TOKEN_ADDRESS,18)`,
+`gasPayingTokenName` should return `Ether` and `gasPayingTokenSymbol` should return `ETH`.
+
+This interface applies to the following contracts:
+
+- `L1StandardBridge`
+- `L2StandardBridge`
+- `L1CrossDomainMessenger`
+- `L2CrossDomainMessenger`
+- `L1Block`
+- `SystemConfig`
```

Comment:
Why this set of contracts? i.e. we can consider deduping to a single source of truth on each chain to reduce the amounts of code/entry points for the gas token functionality
```suggestion
This interface applies to the following contracts:

- `L1Block`
- `SystemConfig`
```